src/analysis/type_search.ml

```diff
@@ -0,0 +1,155 @@
+    | Types.Tconstr (p, args, _) ->
+      let p = Printtyp.rewrite_double_underscore_paths env p in
+      let name = Format.asprintf "%a" Printtyp.path p in
+      Type_parsed.Tycon (name, List.map ~f:aux args)
+    | _ -> Type_parsed.Unhandled
+      in typ |> aux |> Type_expr.normalize_type_parameters
+
+let make_constructible path desc =
+  let holes = match Types.get_desc desc with
+    | Types.Tarrow (l, _, b, _) ->
+      let rec aux acc t =
+        match Types.get_desc t with
+        | Types.Tarrow (l, _, b, _) ->
+          aux (acc ^ with_label l) b
+        | _ -> acc
+      and with_label l =
+        match l with
+        | Ocaml_parsing.Asttypes.Nolabel -> " _"
+        | Labelled s -> " ~" ^ s ^ ":_"
+        | Optional _ -> ""
+      in
+      aux (with_label l) b
+    | _ -> ""
+  in
+  path ^ holes
+
+let make_trie env modules =
+  let rec walk env lident =
+    Env.fold_modules (fun name _ mdl acc ->
+        match mdl.Types.md_type with
+        | Types.Mty_alias _ -> acc
+        | _ ->
+          let lident = Longident.Ldot (lident, name) in
+          T (name, lident, lazy (walk env lident)) :: acc
+      ) (Some lident) env []
+  in
+  List.fold_left
+    ~init:[]
+    ~f:(fun acc name ->
+        let lident = Longident.Lident name in
+        match Env.find_module_by_name lident env with
+        | exception _ -> acc
+        | _ -> T (name, lident, lazy (walk env lident)) :: acc
+      )
+    modules
+
+let doc_to_option = function
+  | `Builtin doc
+  | `Found doc -> Some doc
+  | _ -> None
+
+let compare_result (cost_a, a, _, doc_a, _) (cost_b, b, _, doc_b, _) =
+  let c = Int.compare cost_a cost_b in
+  if Int.equal c 0 then
+    let c = Int.compare (String.length a) (String.length b) in
+    match c, doc_a, doc_b with
+    | 0, Some _, None -> 1
+    | 0, None, Some _ -> -1
+    | 0, Some a, Some b ->
+      Int.compare (String.length a) (String.length b)
+    | _ -> c
+  else c
+
+let run ?(limit = 100) config local_defs comments pos env query trie =
+  let fold_values dir acc =
+    Env.fold_values (fun _ path desc acc ->
+        let open Merlin_sherlodoc in
+        let d = desc.Types.val_type in
+        let typ = type_of env d in
+        let path = Printtyp.rewrite_double_underscore_paths env path in
+        let path = Format.asprintf "%a" Printtyp.path path in
+        let cost = Query_parser.distance_for query ~path typ in
+        if cost >= 1000 then acc
+        else
+          let doc =
+            Locate.get_doc
+              ~config
+              ~env
+              ~local_defs
+              ~comments
+              ~pos
+              (`User_input path)
+            |> doc_to_option
+          in
+          let constructible = make_constructible path d in
+          (cost, path, desc, doc, constructible) :: acc
+      ) dir env acc
+  in
+  let rec walk acc (T (_, dir, children)) =
+    let force () =
+      let _ = Env.find_module_by_name dir env in
+      Lazy.force children
```

Comment:
I've been asked to review and could only find this nitpick: Perhaps the side-effecting `Env.find_module_by_name` function call could be moved to `make_trie`'s `walk` to avoid incorrect access of the lazy `children`? It's not entirely clear to me why a lazy trie needs to be built since it will be fully traversed? The code also looks inspired by `polarity_search.ml` similar trie, so perhaps this way of traversing the environment could be shared?

(Sorry this turned into a chain of thoughts as I'm still discovering Merlin's codebase, feel free to ignore as there could be excellent reasons to do things that way!)

Reply:
Hi @art-w, on https://github.com/ocaml/merlin/pull/1828/commits/87eeea0154bd722bd8b4e35bdff50fc43bafdbf1 I've rewritten the collection of environment values without using the Lazy trie. I haven't noticed any slowdown in use! Thanks for the suggestion!

I haven't changed the polarity search code (because it seems to me that the PR is already big enough ;) )